README.md
# AFB - Autofishbot
Auto fishing bot made in Python 3 for [Virtual Fisher](https://virtualfisher.com/) Discord bot.

## Features
- Auto Fish 
- Menu (TUI)
- Captcha Bypass [*](#captcha-information)
- Full-featured (auto boosts, auto daily, auto sell, and +)
- Interactions (buttons and slash commands)

**Note**: the **experimental-2.0.0** (current) version is still under development, if you are having problems, a stable release (**1.2.1.1**) can be found [here](assets/stable-legacy/stable-1.2.1.1.zip).

## Demo
![demo](assets/images/demo.gif)

## Requirements
- Python [3.10.6](https://www.python.org/downloads/release/python-3106/) or [later versions](https://www.python.org/downloads/)
- [Git](https://git-scm.com/downloads) (optional)

## Installation and Update
(**With** Git) 
- Using the terminal, type:
```bash
git clone https://github.com/thejoabo/autofishbot.git
cd autofishbot
python pip -m install -r requirements.txt
```
(**Without** Git)
- Download the repository [here](https://github.com/thejoabo/autofishbot/archive/refs/heads/main.zip)
- Extract the zip file
- Open a terminal window in the extracted folder and type:
```bash
python pip -m install -r requirements.txt
``` 
## Usage
Using the terminal, **in the repository folder**, type:
```bash
python autofishbot.py 
```
**Notes**:
- You can use ```python autofishbot.py config_name``` to run the bot with a specific config file.
- You can use ```python autofishbot.py --create``` to create a new config file.
- In the first startup, the bot will ask you to fill the configuration file name (which will be stored at '```{autofishbot directory}/configs/```' folder). After setting everything up, run the command again. In case of several config files, you will be prompted to choose one.

## Customization

You can easily customize the options listed below in the automatically generated [config file](assets/template.config):

```conf
#Example
[SYSTEM]
user_token = M@yToke_n123
user_cooldown = 3.5
channel_id = 123456
debug = False

[CAPTCHA]
ocr_api_key = MyK!ey12.3

[NETWORK]
user_agent = 
proxy_ip =
proxy_port =
proxy_auth_user = 
proxy_auth_password = 

[AUTOMATION]
boosts_length = 5
more_fish = True
more_treasures = False
fish_on_exit = True
auto_daily = True
auto_buy_baits = False
auto_sell = True
auto_update_inventory = False

[MENU]
compact_mode = False
refresh_rate = 0.3

[COSMETIC]
pet = dolphin
bait = fish
biome = ocean
```
Detailed information [here](assets/faq.md).


## Captcha Information
All captchas are solved using the OCR.SPACE API for image to text recognition. The reason for choosing an online API was to avoid the annoyance of forcing users to install heavy image recognition modules (and saving the unnecessary effort of creating a specific image recognition model for this type of captcha). Among the API options, the most practical and accessible for the user is OCR.SPACE and, furthermore, it presented a reasonable consistency in correctly identifying the text in the image in the tests performed. **Therefore, to automatically solve the captchas, you will need an [API KEY](#how-do-i-get-my-free-key-)**.

The methodology is quite straight forward, when a new captcha is detected:

1. A request is sent asynchronously to the API for each available OCR engine
2. The answers are filtered to assert those with reasonable certainty 
3. The filtered answers are tested

If all tests fail, a request to regenerate the captcha will be sent (up to 3 times). If the bot, ultimately, fails to solve all captchas, it will wait until you solve it manually.

Keep in mind that the captcha detection method is not flawless, unexpected events can cause some unusual behavior that could influence detection accuracy. Therefore, it should not be left alone without monitoring for longer periods of time.
### Demo (normal)
![captcha-demo](assets/images/captcha-demo.gif)

### Demo (with regen)
![captcha-demo-wrong-code](assets/images/captcha-with-regen.gif)

### HOW DO I GET MY **FREE** KEY ?

All you need to do is use [any](https://temp-mail.io/en) email to get it here: https://ocr.space/ocrapi/freekey.

[Need more help ?](assets/faq.md#ocr_api_key)


## Changelog
### experimental-2.0.0 - 11/20/22 (current)
#### **Added**
- added **Scheduler** class (#38) - experimental
  - Allows slash commands to be sent using keybinds
  - Creates, controls and pseudo-randomizes automation routines (boosts and auto_* features) → attempts to mimic human behavior
  - Sporadically mid-session 'break time'
- **DiscordWrapper** new features
  - added **Proxy** class integration
    - native proxy support (**HTTPS only**)
  - session reconnections
- **Menu** new features
  - **BaseMenu**
    - support for custom keybinds ('*./app/keybinds.json*') by **Keybinder** class
    - CPU and Memory consumption reduced - better logic and structure overall (removed redundant calculations and  function calls, simplification, +)
    - asynchronous notifications 'server' (threaded) → notifications are now queued and have priority and display time
    - 'Config Status' panel split in 'App info' and 'Config Status' (with better choices of information to be displayed)
    - **Popup** widget → extra information (like quests, exotic fishes, current inventory, charms and buffs) can be viewed in a popup format (see example) 
  - **MainMenu(BaseMenu)** - "normal" mode
    - design changes ([see comparison](https://imgur.com/a/tQSEXG3))
    - screen strings are now resized dynamically to fit the terminal size
    - added the Leaderboards panel
    - added more profile information to the Inventory panel
    - better exceptions handling 
  - **CompactMenu(BaseMenu)** - compact mode
    - minor design changes ([see comparison](https://imgur.com/a/tsIcg5O))
    - better exceptions handling 
- **ConfigManager** new features
  - argument support for loading specific configs  (e.g. ```python autofishbot.py some_config_name```) (adaptation of @notvirtio 's solution #51)
  - support for the creation of new (alt) configs using arguments (```python autofishbot.py --create```)
  - new parameters:
    - Network: ```proxy_ip```, ```proxy_port```, ```proxy_auth_user```, ```proxy_auth_password```, ```user_agent```
    - Automation:
      - added ```auto_sell```, ```auto_buy_baits```, ```auto_update_inventory```, ```auto_daily```
      - splitted ```auto_buff``` in ```more_fish``` and ```more_treasures```
      - renamed ```buff_length``` to ```boosts_length```
    - Menu: ```refresh_rate```
    - Cosmetic: ```pet```, ```bait```, ```biome``` (**not yet implemented**)
- **Profile** new features
  - support to charms inventory (*/charms*) + **Charms** datatype class
  - support to user leaderboards (*/pos*) + **Leaderboard** datatype class
  - support to buffs (multipliers list) (*/buffs*) + **Buffs** datatype class
  - support to quests list (*/quests*) + **QuestList** and **Quest** datatype class
#### **Changed**
- refactored **DiscordWrapper** class
  - discord's API (v9 → v10)
  - *connect* function's structure
  - heartbeat routine (adjusting to documentation requirements)
  - *receive_event* function (better error management and pre deserialization)
  - application commands and guild ID are now loaded on start-up
  - *request* function (discord's HTTP API requests):
    - simplification and better data structure
    - better handling of exceptions (and proper returns)
    - better error management to failure requests
    - proper handling of discord's rate limits (429 error) (improvement of @gabeaventh 's solution #36)
- refactored **Captcha** class
  - compartmentalization
  - multi-threading solving (asynchronous ocr-engine requests) → each engine request is made simultaneously, so captcha solving time is significantly faster.
  - captcha detection function (added conditional safeguards)
  - status flags (ensuring transparency of actions for external classes)
- *message_dispatcher* function converted to **Dispatcher** class
  - improvements in conditional gates to send commands (safer method)
  - "**pause**" property, to globally pause the dispatcher (instead of a standalone class)
  - improvements in sending captcha-related commands 
  - improvements in cooldown calculations (to consider for network ping between requests oscillations/processing time, in order to preserve the Gaussian approach of value generation)
- *message_receiver* function converted to **Receiver** class
  - better event checking, parsing and validation
  - simplification (removed unnecessary event validation, improvements in code structure, +)
  - improvements in messages categorization
  - untitled (and unhandled) messages are now properly sanitized and displayed on notifications
  - better logic to detect captchas in raw events
  - incorrect captcha code messages are now treated
  - added of more breakpoints conditionals (in case of abnormal behavior)
- refactored **Profile** class
  - compartmentalization
  - attribute fields (instead of a global fixed list)
  - lists (displayed on menu) are now constructed individually per attribute (datatype classes) using the *list* property
  - *update* function 
  - gold, diamond, emerald and lava fishes are now grouped to **ExoticFish** class
- refactored **CooldownManager** class
  - simplification
  - analysis functions removed
- refactored **Message** class
  - better message parsing and fields categorization
  - untitled and contentless message are now handled
- refactored **ConfigManager** class
  - better loading/saving logic of config files
  - compartmentalization of config parameters (system, captcha, network, automation, menu and cosmetic)
  - better value parsing and validation
- refactored **Debugger** class
  - simplification
  - outputs errors and exceptions to 'debug.log' file instead of displaying them on terminal
- **PauseControl** class removed (integrated in **Dispatcher** class)
-  **MenuManager** class removed (split in **BaseMenu**, **MainMenu(BaseMenu)**, **CompactMenu(BaseMenu)** classes)
- *autoBuff* function removed (integrated in **Scheduler** class)
#### **Fixed**
- fixed a bug in which the *sanitize* function in Message class couldn't parse some characters
- fixed a bug in which  required fields weren't validated in ConfigManager (#48)
- fixed a bug in which captcha.regen counter were redefined after solution failure
- fixed a bug in which the bait name overlaps the border in compact mode - inventory (and related information) can be viewed separately on a popup using their respective keybinds (#50)
- fixed a bug in which the curses kept blinking on screen's last update position
- 
...

[FULL CHANGELOG HERE](assets/changelog.md)


## Contributing
- Pull requests are welcome. For major changes, open an issue first to discuss what you would like to change.
- If you notice any errors, bugs or strange behavior **[PLEASE OPEN AN ISSUE](https://github.com/thejoabo/virtualfisher-bot/issues/new)** containing a screenshot or describing the problem.
- [**Suggestions**](https://github.com/thejoabo/virtualfisher-bot/discussions/new) and [**Pull requests**](https://github.com/thejoabo/virtualfisher-bot/pulls) are welcome.



assets/changelog.md
# Changelog
<!-- ## v0.0.0 - x/x/22
### **Added**
### **Changed**
### **Fixed** -->
## experimental-2.0.0 - 11/20/22 (current)
#### **Added**
- added **Scheduler** class ([#38]([#38](https://github.com/thejoabo/autofishbot/issues/38))) - experimental
  - Allows slash commands to be sent using keybinds
  - Creates, controls and pseudo-randomizes automation routines (boosts and auto_* features) → attempts to mimic human behavior
  - Sporadically mid-session 'break time'
- **DiscordWrapper** new features
  - added **Proxy** class integration
    - native proxy support (**HTTPS only**)
  - session reconnections
- **Menu** new features
  - **BaseMenu**
    - support for custom keybinds ('*./app/keybinds.json*') by **Keybinder** class
    - CPU and Memory consumption reduced - better logic and structure overall (removed redundant calculations and  function calls, simplification, +)
    - asynchronous notifications 'server' (threaded) → notifications are now queued and have priority and display time
    - 'Config Status' panel split in 'App info' and 'Config Status' (with better choices of information to be displayed)
    - **Popup** widget → extra information (like quests, exotic fishes, current inventory, charms and buffs) can be viewed in a popup format (see example) 
  - **MainMenu(BaseMenu)** - "normal" mode
    - design changes ([see comparison](https://imgur.com/a/tQSEXG3))
    - screen strings are now resized dynamically to fit the terminal size
    - added the Leaderboards panel
    - added more profile information to the Inventory panel
    - better exceptions handling 
  - **CompactMenu(BaseMenu)** - compact mode
    - minor design changes ([see comparison](https://imgur.com/a/tsIcg5O))
    - better exceptions handling 
- **ConfigManager** new features
  - argument support for loading specific configs  (e.g. ```python autofishbot.py some_config_name```) (adaptation of @notvirtio 's solution [#51](https://github.com/thejoabo/autofishbot/pull/51))
  - support for the creation of new (alt) configs using arguments (```python autofishbot.py --create```)
  - new parameters:
    - Network: ```proxy_ip```, ```proxy_port```, ```proxy_auth_user```, ```proxy_auth_password```, ```user_agent```
    - Automation:
      - added ```auto_sell```, ```auto_buy_baits```, ```auto_update_inventory```, ```auto_daily```
      - splitted ```auto_buff``` in ```more_fish``` and ```more_treasures```
      - renamed ```buff_length``` to ```boosts_length```
    - Menu: ```refresh_rate```
    - Cosmetic: ```pet```, ```bait```, ```biome``` (**not yet implemented**)
- **Profile** new features
  - support to charms inventory (*/charms*) + **Charms** datatype class
  - support to user leaderboards (*/pos*) + **Leaderboard** datatype class
  - support to buffs (multipliers list) (*/buffs*) + **Buffs** datatype class
  - support to quests list (*/quests*) + **QuestList** and **Quest** datatype class
#### **Changed**
- refactored **DiscordWrapper** class
  - discord's API (v9 → v10)
  - *connect* function's structure
  - heartbeat routine (adjusting to documentation requirements)
  - *receive_event* function (better error management and pre deserialization)
  - application commands and guild ID are now loaded on start-up
  - *request* function (discord's HTTP API requests):
    - simplification and better data structure
    - better handling of exceptions (and proper returns)
    - better error management to failure requests
    - proper handling of discord's rate limits (429 error) (improvement of @gabeaventh 's solution [#36](https://github.com/thejoabo/autofishbot/pull/36))
- refactored **Captcha** class
  - compartmentalization
  - multi-threading solving (asynchronous ocr-engine requests) → each engine request is made simultaneously, so captcha solving time is significantly faster.
  - captcha detection function (added conditional safeguards)
  - status flags (ensuring transparency of actions for external classes)
- *message_dispatcher* function converted to **Dispatcher** class
  - improvements in conditional gates to send commands (safer method)
  - "**pause**" property, to globally pause the dispatcher (instead of a standalone class)
  - improvements in sending captcha-related commands 
  - improvements in cooldown calculations (to consider for network ping between requests oscillations/processing time, in order to preserve the Gaussian approach of value generation)
- *message_receiver* function converted to **Receiver** class
  - better event checking, parsing and validation
  - simplification (removed unnecessary event validation, improvements in code structure, +)
  - improvements in messages categorization
  - untitled (and unhandled) messages are now properly sanitized and displayed on notifications
  - better logic to detect captchas in raw events
  - incorrect captcha code messages are now treated
  - added of more breakpoints conditionals (in case of abnormal behavior)
- refactored **Profile** class
  - compartmentalization
  - attribute fields (instead of a global fixed list)
  - lists (displayed on menu) are now constructed individually per attribute (datatype classes) using the *list* property
  - *update* function 
  - gold, diamond, emerald and lava fishes are now grouped to **ExoticFish** class
- refactored **CooldownManager** class
  - simplification
  - analysis functions removed
- refactored **Message** class
  - better message parsing and fields categorization
  - untitled and contentless message are now handled
- refactored **ConfigManager** class
  - better loading/saving logic of config files
  - compartmentalization of config parameters (system, captcha, network, automation, menu and cosmetic)
  - better value parsing and validation
- refactored **Debugger** class
  - simplification
  - outputs errors and exceptions to 'debug.log' file instead of displaying them on terminal

app/config.py
#------------------------ IMPORTS --------------------------#
from __future__ import annotations

from . import *
from configparser import ConfigParser
from webbrowser import Error as wbError, open as wbOpen
from datetime import date
from random import randbytes
from os import listdir, getcwd
from re import sub 
from sys import argv


#------------------------ CONSTANTS --------------------------#
REPO_CONFIG = 'https://github.com/thejoabo/virtualfisher-bot/blob/main/assets/template.config'


#------------------------- CLASSES ---------------------------#
class GenericException(Exception): pass

class MissingRequiredFieldError(Exception): 
    def __init__(self, field_name: str) -> None:
        super().__init__(f"The field '{field_name}' is required.")
        
class OutsideBoundariesError(Exception):
    def __init__(self, limits: tuple, value: any, field: str = 'Unk') -> None:
        super().__init__(f"{field}: Value outside boundaries ({limits[0]}, {limits[1]}), '{value}' given.") 
        
@dataclass(slots=True)
class ConfigManager:
    '''The configuration file manager to autofisherbot.'''
    #Todo: use argparse library instead of sys.argv for better (and more featured) argument parsing
    #Todo: ^^ https://docs.python.org/3/library/argparse.html#module-argparse
    
    
    #System
    user_token: str = ''
    user_cooldown: float = 3.5
    channel_id: str = ''
    debug: bool = False
    config_path: str = './configs/'
    
    #Captcha
    ocr_api_key: str = ''
    #manual_mode: bool = False
    
    #Network
    user_agent: str = ''
    proxy_ip: str = ''
    proxy_port: str = ''
    proxy_auth_user: str = ''
    proxy_auth_password: str = ''
    
    #Automation
    boosts_length: int = 5
    more_fish: bool = False
    more_treasures: bool = False
    fish_on_exit: bool = False
    auto_daily: bool = False
    auto_buy_baits: bool = False
    auto_sell: bool = False
    auto_update_inventory: bool = False
    
    #Menu
    compact_mode: bool = False
    refresh_rate: float = 0.3
    #number_formatting: str = 'default'
    
    #Cosmetic
    pet: str = ''
    bait: str = ''
    biome: str = ''
    
    _path: str = field(init=False, repr=False) 
    _configs: list[str] = field(init=False, repr=False)
    
    def __post_init__(self) -> None:
        self._path = f'{getcwd()}/{self.config_path}'
        self._configs = self.list_configs()
        
        
        if len(argv) >= 2:
            if argv[1] == '--create':
                self.create_config()
                exit()
        if self._configs != []:
            if len(self._configs) > 1:
                selected = self.choice_dialog()
                self.load_config(selected)
            else:
                self.load_config(self._configs[0])
        else:
            self.create_config()  
    
    def load_config(self, selected_config: str) -> None:
        '''Loads a config file.'''
        print(f'[*] Loading "{selected_config}" ...')
        
        cfg = ConfigParser()
        cfg.read(f'{self._path}/{selected_config}')
        
        try:
            system = cfg['SYSTEM']
            captcha = cfg['CAPTCHA']
            network = cfg['NETWORK']
            automation = cfg['AUTOMATION']
            menu = cfg['MENU']
            cosmetics = cfg['COSMETIC']
        except KeyError as e:
            self.err_dialog(f'[E] Outdated config file.\nTIP: You can use "python autofishbot.py --create" to create a new one at any time.')
            exit(1)
            
        try:
            #System
            self.user_token = self.to_str(system['user_token'], field='USER_TOKEN')
            self.user_cooldown = self.to_float(system['user_cooldown'], field='USER_COOLDOWN')
            self.channel_id = self.to_str(system['channel_id'], field='CHANNEL_ID')
            self.debug = self.to_bool(system['debug'])
            
            #Captcha
            self.ocr_api_key = self.to_str(captcha['ocr_api_key'], field='OCR_API_KEY')
            
            #Network
            self.user_agent = self.to_str(network['user_agent'], required=False)
            self.proxy_ip = self.to_str(network['proxy_ip'], required=False)
            self.proxy_port = self.to_str(network['proxy_port'], required=False)
            self.proxy_auth_user = self.to_str(network['proxy_auth_user'], required=False)
            self.proxy_auth_password = self.to_str(network['proxy_auth_password'], required=False)
            
            #Automation
            self.boosts_length = self.to_int(automation['boosts_length'], field='BOOSTS_LENGTH')
            self.more_fish = self.to_bool(automation['more_fish'])
            self.more_treasures = self.to_bool(automation['more_treasures'])
            self.fish_on_exit = self.to_bool(automation['fish_on_exit'])
            self.auto_daily = self.to_bool(automation['auto_daily'])
            self.auto_buy_baits = self.to_bool(automation['auto_buy_baits'])
            self.auto_sell = self.to_bool(automation['auto_sell'])
            self.auto_update_inventory = self.to_bool(automation['auto_update_inventory'])
            
            #Menu
            self.compact_mode = self.to_bool(menu['compact_mode'])
            self.refresh_rate = self.to_float(menu['refresh_rate'], field='REFRESH_RATE', bd=(0.1, 1))
            
            #Cosmetic
            self.pet = self.to_str(cosmetics['pet'], False)
            self.bait = self.to_str(cosmetics['bait'], False)
            self.biome = self.to_str(cosmetics['biome'], False)
        except KeyError as e:
            self.err_dialog(f'[E] Err -> {e} key missing | Outdated config file.')
        except (MissingRequiredFieldError, 
                OutsideBoundariesError, 
                GenericException, 
                ValueError) as e:
            exit(f'[E] Err -> {e}')
            
    def create_config(self) -> None:
        '''Creates a new config file.'''
        print(f'[*] Generating new config...')
        
        cfg = ConfigParser()
        cfg['SYSTEM'] = {
            'user_token': self.user_token,
            'user_cooldown': self.user_cooldown,
            'channel_id': self.channel_id,
            'debug': self.debug
        }
        
        cfg['CAPTCHA'] = {
            'ocr_api_key': self.ocr_api_key
        }
        
        cfg['NETWORK'] = {
            'user_agent': self.user_agent,
            'proxy_ip': self.proxy_ip,
            'proxy_port': self.proxy_port,
            'proxy_auth_user': self.proxy_auth_user,
            'proxy_auth_password': self.proxy_auth_password
        }
        
        cfg['AUTOMATION'] = {
            'boosts_length': self.boosts_length,
            'more_fish': self.more_fish,
            'more_treasures': self.more_treasures,
            'fish_on_exit': self.fish_on_exit,
            'auto_daily': self.auto_daily,
            'auto_buy_baits': self.auto_buy_baits,
            'auto_sell': self.auto_sell,
            'auto_update_inventory': self.auto_update_inventory
        }
        
        cfg['MENU'] = {
            'compact_mode': self.compact_mode,
            'refresh_rate': self.refresh_rate
        }
        
        cfg['COSMETIC'] = {
            'pet': self.pet,
            'bait': self.bait,
            'biome': self.biome
        }

        try:
            cfg_name = input(f'[?] Please insert a name for this config: ')
            if not cfg_name or len(sub('[a-zA-Z0-9 ]', '', cfg_name)) > 0:
                print('[!] Invalid name, a new name will be used instead.')
                cfg_name = self.make_name()
        except KeyboardInterrupt:
            exit(f'\n[!] User exited.')

        try:
            cfg_name += '.config'
            with open(f'{self._path}/{cfg_name}' , 'w') as f:
                cfg.write(f)
            exit_message = f'\n[!] Config "{cfg_name}" successfully created at "{self._path}".\n[!] Please fill in the configuration information...'
            wbOpen(f'{self._path}/{cfg_name}' )
        except wbError as e:
            exit_message = f'\n[E] Unable to open default text editor, please edit it manually. Details: {e}'
        except Exception as e:
            exit_message =f'\n[E] Unable to create new config -> {e}\n[!] NOTE: You can download a sample config file at: {REPO_CONFIG}'
        finally:
            print(exit_message)
        
    def list_configs(self, arr: list = []) -> list:
        '''List all .config files from a directory.'''
        for file in listdir(self._path):
            if file.find('.config') > -1:
                arr.append(file)
        return arr

    def to_str(self, value: any, required: bool = True, field: str = None) -> str:
        '''Validate string values.'''
        value = sub('["\']', '', value)
        if value:
            return str(value)
        else: 
            if required:
                raise MissingRequiredFieldError(field)
            else:
                return None
            
    def to_bool(self, value: any) -> bool:
        '''Converts and evaluates string values to boolean.'''
        if value:
            return True if str(value).lower() in ['1', 'true', 'on'] else False
        else:
            return False

    def to_int(self, value: any, field: str = None) -> int:
        '''Converts and evaluates string values to integer.'''
        if value:
            try:
                if int(value) > 12.5:
                    return 20
            except ValueError as e:
                raise GenericException(f'{field}: {e}')
        return 5

    def to_float(self, value: any, field: str = None, bd: tuple = (2, 3.5)) -> float:
        '''Converts and evaluates string values to float.'''
        if value:
            try:
                value = float(value)
            except ValueError as e:
                raise GenericException(f'{field}: {e}.')
            
            if self.compare(value, bd):
                return float(value)
            else:
                raise OutsideBoundariesError((2, 3.5), value, field)
        else:
            if field == 'REFRESH_RATE':
                return 0.3
            else:
                raise MissingRequiredFieldError(field)

    def compare(self, value: float, bd: tuple) -> bool:
        '''Compares if value is between boundaries.'''
        l_min, l_max = bd
        if value >= l_min and value <= l_max:
            return True
        return False
    
    def err_dialog(self, message: str) -> None:
        '''Error dialog to create a new config.'''
        print(f'{message}')
        try:
            if input(f'[?] Do you want to create a new config ? (y/n) ').lower() == 'y':
                self.create_config()
            else: 
                exit(f'[!] User exited.')
        except KeyboardInterrupt: 
            exit(f'\n[!] User exited.')
        except Exception as e: 
            #debugger.debug(e, 'Exception')
            exit(f'\n[!] User exited ({e}).')
    
    def choice_dialog(self) -> str:
        '''Choosing dialog to select one config.'''
        if len(argv) >= 2:
            try: 
                index = int(argv[1])
                if index > 0 and index <= len(self._configs):
                    return self._configs[index - 1]
                else:
                    print(f'Invalid choice. {index} is outside the scope.')
            except ValueError:
                for name in self._configs:
                    if name.lower() == f'{argv[1].lower()}.config'or name.lower() == argv[1].lower():
                        return name
                print(f'Invalid choice. "{argv[1]}" is outside the scope.')
        
        print(f'[!] Multiple configs detected in {self._path} folder.\nPlease choose:')
        for k, config in enumerate(self._configs):
            print(f'\t{k + 1} - "{config}"')
        while True:
            try:
                index = int(input(f'[?] Insert the config number: '))
                if index > 0 and index <= len(self._configs):
                    return self._configs[index - 1]
                else: 
                    raise GenericException(f'Invalid choice. {index} is outside the scope.')
            except KeyboardInterrupt:
                exit('\n[!] User exited.')
            except Exception as e:
                print(f'[E] Try again. Err -> {e}')
    
    def make_name(self) -> str:
        return f'{date.today().strftime("%d-%m-%y")}-{randbytes(5).hex()}' 
    


# --------- INIT ---------#
if __name__ == "__main__":
    config = ConfigManager()
    print(config)
